src/mdoc_zk/layout.rs

```diff
@@ -0,0 +1,451 @@
+    ///
+    /// Panics if the input slice is not of the length given by [`Self::signature_input_length()`].
+    pub(super) fn split_signature_input<'a>(
+        &self,
+        input: &'a mut [FieldP256],
+    ) -> SplitSignatureInput<'a> {
+        assert_eq!(input.len(), self.signature_input_length());
+        // After this assertion, all subsequent `unwrap()` and `split_at_mut()` calls should not panic.
+
+        let (implicit_one, input) = input.split_first_mut().unwrap();
+        let (issuer_public_key_x, input) = input.split_first_mut().unwrap();
+        let (issuer_public_key_y, input) = input.split_first_mut().unwrap();
+        let (e_session_transcript, input) = input.split_first_mut().unwrap();
+        let (mac_tags, input) = input.split_at_mut(3 * 2 * 128);
+        let (mac_verifier_key_share, input) = input.split_at_mut(128);
+        let (e_credential, input) = input.split_first_mut().unwrap();
+        let (device_public_key_x, input) = input.split_first_mut().unwrap();
+        let (device_public_key_y, input) = input.split_first_mut().unwrap();
+        let (credential_ecdsa_witness, input) = input.split_at_mut(EcdsaWitness::LENGTH);
+        let (device_ecdsa_witness, input) = input.split_at_mut(EcdsaWitness::LENGTH);
+        let (mac_witnesses, input) = input.split_at_mut(3 * 256 * 2 / 2);
+        assert!(input.is_empty());
+
+        SplitSignatureInput {
+            implicit_one,
+            issuer_public_key_x,
+            issuer_public_key_y,
+            e_session_transcript,
+            mac_tags: mac_tags.try_into().unwrap(),
+            mac_verifier_key_share: mac_verifier_key_share.try_into().unwrap(),
+            e_credential,
+            device_public_key_x,
+            device_public_key_y,
+            credential_ecdsa_witness: EcdsaWitness::new(
+                credential_ecdsa_witness.try_into().unwrap(),
+            ),
+            device_ecdsa_witness: EcdsaWitness::new(device_ecdsa_witness.try_into().unwrap()),
+            mac_witnesses: mac_witnesses.try_into().unwrap(),
+        }
+    }
+
+    /// Returns the length of the input for the hash circuit, in field elements.
+    ///
+    /// This includes all public and private inputs, including the implicit 1.
+    pub(super) fn hash_input_length(&self) -> usize {
+        match self.version {
+            CircuitVersion::V6 => {}
+        }
+        1 // implicit 1
+            + usize::from(self.attributes) * AttributeInput::LENGTH // attribute CBOR data
+            + 20 * 8 // time in RFC 3339 format
+            + 3 * 2 // MAC tags
+            + 1 // MAC verifier key share
+            + 256 // hash of credential
+            + 2 * 256 // device public key
+            + 8 // number of SHA-256 blocks
+            + SHA_256_INPUT_WIRES // padded SHA-256 input for credential
+            + SHA_256_CREDENTIAL_MAX_BLOCKS * Sha256BlockWitness::LENGTH // SHA-256 witness for credential
+            + 12 // validFrom CBOR offset
```

Comment:
It takes 12 GF(2^128) elements to represent the offset into a CBOR document? That seems like a lot.